Fixed bug with Russound integration and added 2 way state support. (Bumped up to v0.1.7)

acambitsis opened the pull request:
**Description:**
Fixed bug arising from not supporting TURN_ON state (fixes issue https://github.com/home-assistant/home-assistant/issues/5012)
Implemented state support in 0.1.7 such that component state is returned from the actual AMP. (Still uses polling model though).
Tested it with home-assistant users @laf (original developer of the module) and @hofsta.  Works fine with their Russounds.

fixes https://github.com/home-assistant/home-assistant/issues/5012

acambitsis committed:
Bumped up version to use 0.1.7 of Russound integration module.
Fixed bug arising from not supporting TURN_ON state (fixes issue https://github.com/home-assistant/home-assistant/issues/5012)
Implemented state support in 0.1.7 such that component state is returned from the actual AMP. (Still uses polling model though).
Tested it with home-assitant users @laf (original developer of the module) and @hofsta.  Works fine with their Russounds.

homeassistant commented:
Hi @acambitsis, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5756).

Once you do that we will be able to review and accept this pull request.

Thanks!

acambitsis committed:
Made styling / compliance changes and updated correct version of russound module on requirements_all.txt.

homeassistant commented:
Hi @acambitsis, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5756).

Once you do that we will be able to review and accept this pull request.

Thanks!

acambitsis commented:
I have now signed the CLA.

laf commented:
I can confirm this works as expected, tested on a Russound Cav 6.6 unit. Adds support for two way communication so is a huge +.

pvizeli reviewed (changes requested):
Please don't make IO on propertys. Use update() to cache all stuff like volume or sorce and only return it in property

acambitsis commented:
@pvizeli happy to fix whatever is needed.  I assume the issue is within the russound_net.py which is directly part of the PR and not on the dependent module (https://github.com/laf/russound/blob/master/russound/russound.py)?
I need a bit more assistance on what is needed to be fixed.  Can you perhaps point me to one or two lines of code where the issue is happening (e.g. is line 116?).
Thanks,
Andreas.

armills commented on homeassistant/components/media_player/russound_rnet.py:
```diff
@@ -76,13 +75,16 @@ class RussoundRNETDevice(MediaPlayerDevice):
+
+        Note: now uses the Russound device directly to obtain the state and
+        volume level, and therefore we no longer
+        have properties in this class for these to variables.
```
This comment block can be removed. It's good to note in the PR, but after it gets merged, the comment won't make sense any more.

acambitsis replied:
Agree and implemented.

armills commented on homeassistant/components/media_player/russound_rnet.py:
```diff
@@ -91,33 +93,54 @@ def name(self):
-        """Return the state of the device."""
-        return self._state
+        """Return the state of the device. directly from the device."""
+        if self._russ.get_power('1', self._zone_id) == 0:
```
As pvizeli mentioned, no IO can be performed in the entity properties. Instead you need to create an update method, and store the state in `self._state`, to return here. Same for all other properties.

[There's an example here that should help](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/media_player/nad.py#L112), or you can look at other platforms for more examples.

acambitsis replied:
Very useful info.  Thank you.  The NAD.py is a great example as its simple to follow, but demonstrates the correct implementation.
I have implemented the changes.

armills commented on homeassistant/components/media_player/russound_rnet.py:
```diff
@@ -91,33 +93,54 @@ def name(self):
+        try:
+            return self._sources[index]
+        except IndexError:
+            return self._sources[0]
```
Returning `None` here might be better, to indicate that we don't know what source is selected.

acambitsis replied:
Agree and implemented.

emlove reviewed (changes requested):
Nice PR! Getting real status from the device will be a big improvement. I tried to clarify pvizeli's request a bit, and added a few extra notes.

acambitsis commented:
@armills many thanks for the detailed feedback.  I should be able to address this sometime in the next 24 hours.

pvizeli commented:
https://github.com/home-assistant/home-assistant/issues/4210

acambitsis commented:
@armills and @pvizeli thanks for the follow-up info.  It was super helpful,  I have made the changes.  I will test tonight and if all is good I will check-in.  One question re the following though:
Currently the mute_volume method is implemented and the config is set to support SUPPORT_VOLUME_MUTE.  However the integration into Russound (russopund.py) doesn't currently return the mute state.  (It is possible but this was never implemented and I don't have time at the moment to sort that out).  My question therefore is should I leave the mute functionality as is (i.e. it allows you to toggle mute but will not return the mute state), or is it better to remove SUPPORT_VOLUME_MUTE?

armills commented:
It is OK to implement mute functionality without getting the current Mute state. However, there is a problem with the current mute functionality that it isn't setting the mute state to the requested state, it's just toggling it whenever the service is called.

I would just leave it in for now, since it's been released and people could be using it. If there are separate commands for "Mute ON" vs "Mute OFF" that would be better, and actual mute state would be best. That can be a separate PR though.

acambitsis committed:
Changed handling of properties to be compliant with https://github.com/home-assistant/home-assistant/issues/4210
(Specifcailly added member variables for state, volume and source to cache these values, and introduced Update() method to set their values).

Now returns None if the selected source index that is returned from russound is greater than the length of the specified source list in the yaml config.
Removed unnecesary comment.

acambitsis commented:
@armills and @pvizeli  thanks for the super useful info.  This was my first contribution to HA and I appreciate all the guidance.  I have checked in the changes - hopefully I haven't missed anything.  I have tested it on my side and all functions work as expected.

laf commented:
Thanks @acambitsis. I'll give this a whirl over the next few days :)

acambitsis committed:
Removed blank line after docstring.

armills commented on homeassistant/components/media_player/russound_rnet.py:
```diff
@@ -91,7 +114,7 @@ def name(self):
     @property
     def state(self):
-        """Return the state of the device."""
+        """Return the state of the device. directly from the device."""
```
"directly from the device" can be dropped.

acambitsis replied:
Well spotted. My bad.

armills commented on homeassistant/components/media_player/russound_rnet.py:
```diff
@@ -79,10 +78,34 @@ def __init__(self, hass, russ, sources, zone_id, extra):
+        self._volume = None
+        self._source = None
+
+        self.update()
```
Instead of calling `self.update()` at the end of `__init__`, add `True` to the end of `add_devices` here: https://github.com/home-assistant/home-assistant/pull/5756/files#diff-94a55579c81fd6ed2ecfc6300f04aca5R68

Example: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/switch/dlink.py#L56

There's still a lot of devices that are doing this the old way (including nad that I linked), but this is the preferred new way to do the initial update.

acambitsis replied:
Change made and will checkin now.

armills replied:
Don't forget to drop the `self.update()` :)

emlove reviewed (changes requested):
Just two more small changes, and this should be good to go. Thanks for following up and addressing everything!

acambitsis committed:
Removed updated() in class init and added True paramter to add_devices in setup_platform.

acambitsis committed:
Dropped the no longer needed self.update()

armills commented:
Nice work. Thanks for your contribution! 🌸

acambitsis commented:
It's a pleasure. Hope to get an opportunity to contribute more in future. HA is really awesome software and it's a great community. Thanks for the patience with the niggles on my first contribution.

partytimeexcellent commented:
@acambitsis. Great job on this! You really made some huge improvements here and my system is fundamentally better because of it. Cheers!

acambitsis commented:
Great to hear @partytimeexcellent. Thanks for the message. It's great to get that kind of feedback. Especially since it's my first contribution not just to HA but to open source in general.

acambitsis commented:
Hey @armills. I see that this PR has been released with 0.38 of HA which is great.  I noticed that there was nothing mentioned about the Russound component being improved in the [blog post](https://home-assistant.io/blog/2017/02/11/alert-appletv-mqtt-yeelight/).  Anything I should have done to ensure that this was incorporated in the change log?  Whats the best way to let the community know of these fixes / improvements?

armills commented:
Right now, the devs try to scan through the commit log between releases to write up the blog post, but they encourage contributors to comment on the [release PR](https://github.com/home-assistant/home-assistant/pull/5833) when they open it up to add notes there. For example, you can see the yeelight notes were added. I'm not sure this is really documented anywhere, since this practice has just started as Home Assistant has been growing.

For right now, it might be best to ask in the [devs chat](https://gitter.im/home-assistant/home-assistant/devs). I'm not sure what the procedure is for features that got missed in the initial blog post.
